### PR TITLE
#658 Change upstream Regex to account for blank placeholders

### LINKED ISSUE #658: Routing question: placeholders vs slash
## Expected Behavior

* Request to `/api/` would go to `/` downstream
* Request to `/api/test` would go to `/test` downstream

## Actual Behavior / Motivation for New Feature

* Reqeust to `/api/` is not resolved (see error below)
* Request to `/api/test` goes to downstream `/test

## Steps to Reproduce the Problem

  1. Have something listening on localhost:8000
  1. Configure ocelot to listen on localhost:5000
  1. Use config below

Config:
```json
{
    "ReRoutes": [
        {
            "DownstreamPathTemplate": "/{everything}",
            "DownstreamScheme": "http",
            "DownstreamHostAndPorts": [
                {
                    "Host": "localhost",
                    "Port": 8000
                }
            ],
            "UpstreamPathTemplate": "/api/{everything}",
            "UpstreamHttpMethod": [
                "GET",
                "POST"
            ]
        }
    ],
    "GlobalConfiguration": {
        "BaseUrl": "http://localhost:5000"
    }
}
```

## Specifications

  - Version: 12.0.1
  - Platform: mac
  - Subsystem: ?

## Comments

It is interesting to note, that if I configure ReRoute as `/{everything}` -> `/{everything}`, then it works, so it seems to me that `{everything}` wildcard match is not consistent in theese two cases.

### PR DESCRIPTION
**Pull request by AndyConlisk**
## Fixes #658
- #658 

## Proposed Changes

  - To fix the issue mentioned in #658 this adds an additional `Regex` and to account for blank last placeholder.

### CODE REVIEW

**Comment by raman-m** [comment]
@AndyConlisk 
Thanks for the PR!
The [feature branch](/AndyConlisk/Ocelot/tree/master) (**master**) has been [rebased](#event-10015455639) onto [ThreeMammals:develop](../tree/develop)!
And we have the green build.
So, we can go with code review now...

---

I don't see **develop** branch in [your fork](https://github.com/DanHarltey/Ocelot/branches)! Your fork is too old.

Could you [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) please? So, new **develop** branch will occur with all top commits!

Could you [add me as collaborator](/AndyConlisk/Ocelot/settings/access) to your forked repo please? I will create **develop** branch and make it default.

**Comment by raman-m** [comment]
@skyflyer Hi Miha!
As an issue author,
Could you review the code of this solution please?

**Comment by AndyConlisk** [comment]
> @AndyConlisk Thanks for the PR! The [feature branch](/AndyConlisk/Ocelot/tree/master) (**master**) has been [rebased](#event-10015455639) onto [ThreeMammals:develop](../tree/develop)! And we have the green build. So, we can go with code review now...
> 
> I don't see **develop** branch in [your fork](https://github.com/DanHarltey/Ocelot/branches)! Your fork is too old.
> 
> Could you [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) please? So, new **develop** branch will occur with all top commits!
> 
> Could you [add me as collaborator](/AndyConlisk/Ocelot/settings/access) to your forked repo please? I will create **develop** branch and make it default.

@raman-m

I added you as a collaborator to my fork. Please let me know if there is anything else you need me to do.

Thank you

**Comment by skyflyer** [comment]
> @skyflyer Hi Miha!
> As an issue author,
> Could you review the code of this solution please?

Hello @raman-m!

Unfortunately, I'm not really acquainted with the code base and won't be able to either reproduce the error or test the fix right now. Sorry about that.

But, I will have a look at reproducing the scenario later, if that helps.

**Comment by skyflyer** [comment]
So, I'm not sure if my rationale is sound, but this does not seem to work, at least not for the case I reported in the issue.

First, do understand, that we're not using Ocelot anymore (the project that was the consumer of this library unfortunately died some time ago). Also understand that I'm not in a position to really review the code, as I'm not really acquainted with the codebase nor the roadmap and the design of the library. So take the following comments with that in mind. :smile:

The general idea was that ocelot lib would route requests coming to `/api` to some services running without the `/api` prefix. In other words, requests coming to http://localhost:5000/api would be routed (or proxied would be a better term, I think) to http://localhost:8000/ and by that logic, requests coming to http://localhost:5000/api/foo would be routed to http://localhost:8000/api/foo.

In order to verify this behaviour, I cloned the develop branch from the main repo and "squash merged" the PR 1333 into that, so the changes are applied on top of the latest develop bits.

I then used a simple config (see below) and tried to request the resources. I've setup python's simple http server (`python3 -m http.server 8000`) in the parent directory where `Ocelot` (the repository) and `OcelotDemo` (the test program) were hosted.

1. request to http://localhost/api works ✅
2. request to http://localhost/api/ does not work ❌
    * error message: Failed to match Route configuration for upstream path: /api/, verb: GET.
4. request to http://localhost/api/foo works ✅
5. request to http://localhost/api/OcelotDemo/ocelot.json works ✅


<details><summary>Simple config</summary>

```json
{
    "Routes": [
        {
        "DownstreamPathTemplate": "/{id}",
        "DownstreamScheme": "http",
        "DownstreamHostAndPorts": [
            {
                "Host": "localhost",
                "Port": 8000
            }
        ],
        "UpstreamPathTemplate": "/api/{id}",
        "UpstreamHttpMethod": [ "Get", "Post" ]
        }
    ],
    "GlobalConfiguration": {
        "BaseUrl": "https://localhost:5000"
    }
}
```
</details>

**Review by RaynaldM — APPROVED** [review]

**Comment by raman-m** [comment]
> @AndyConlisk commented [on Aug 5](#issuecomment-1666552607)

Thanks!
All that's left to do is making it default. 👇 

image

**Comment by AndyConlisk** [comment]
@raman-m I have made the develop branch to be the default branch.

**Comment by raman-m** [comment]
@skyflyer commented [on Aug 6](#issuecomment-1666866611):
> 2. request to http://localhost/api/ does not work ❌
    * error message: Failed to match Route configuration for upstream path: /api/, verb: GET.

Thanks for user scenarios.
Yeah, we will test this scenario too.
And, on my opinion, all user's scenarios from your issue should be tested, but they are not, based on current source code.

**Comment by raman-m** [comment]
> @RaynaldM approved these changes [on Aug 7](#pullrequestreview-1564708272)

Ray, why did you approve the PR?
There are no unit tests at all! 🤣 
Even a beautiful code without tests should not be approved. 😃

File: src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs [hunk 1]
```diff
@@ -70,6 +70,29 @@ public Response<List<PlaceholderNameAndValue>> Find(string path, string query, s
...
+                    {
+                        delimiter = '&';
+                        nextDelimiter = '&';
+                    }
+
+                    //should_find_multiple_query_string_and_path makes test pass
+                    if (NotPassedQueryString(pathTemplate, counterForTemplate) && NoMoreForwardSlash(pathTemplate, counterForTemplate))
+                    {
+                        delimiter = '?';
+                        nextDelimiter = '?';
```
Comment (raman-m) [inline src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs:86]:
```suggestion
                        delimiter = nextDelimiter = '?';
```

File: src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs [hunk 2]
```diff
@@ -70,6 +70,29 @@ public Response<List<PlaceholderNameAndValue>> Find(string path, string query, s
 
                     counterForTemplate = endOfPlaceholder;
                 }
+                else if (TemplateDoesNotEndInForwardSlash(pathTemplate) && IsLastPlaceholder(path, counterForPath, pathTemplate))
+                {
```
Comment (raman-m) [inline src/Ocelot/DownstreamRouteFinder/UrlMatcher/UrlPathPlaceholderNameAndValueFinder.cs:74]: The code for this **if**-block was copied from lines 26-46.
Based on the function of code snippet, why not to define private helper-method to be reused in both places?

**Review by raman-m — CHANGES_REQUESTED** [review]
@AndyConlisk 
Although the code looks solid, but it can be improved. 👇 

---
### ${\color{orange}There\ are\ no\ new\ tests!}$
The new code blocks are not covered by unit tests.
There are no unit tests for user scenarios from the linked issue.
Also, there are no acceptance tests to make sure Ocelot pipeline will handle all user scenarios correctly.

**Comment by raman-m** [comment]
### Related to
- #145 
- #158 
- 9ba57f8ba654f66d5515fb96b7500ec67bf846c7

**Comment by raman-m** [comment]
@AndyConlisk Do you have time to fix [code review issues](#pullrequestreview-1577061541)?
Do you have an intention to write unit tests at all?

**Comment by AndyConlisk** [comment]
I don't know when I will have time to write the new unit tests. I can try in the near future, but I can't give a solid commitment at this time.

**Comment by raman-m** [comment]
@AndyConlisk commented [on Nov 7](#issuecomment-1797062176)

Lazy bones! :wink:

**Comment by raman-m** [comment]
Duplicate of #1911

**Comment by raman-m** [comment]
### Original issue #658 was fixed by #1911

**Pull request closed without merging**